### ISSUE TITLE
Contracts on `self`

### ISSUE THREAD

**Issue by shockone** [egonSchiele/contracts.ruby#191]
Let's say I have the following contracts

```ruby
Contract Alkohol, Underaged => String
def sell(drink, person)
  "Sorry, come back in #{18 - person.age} years."
end

Contract Drink, Person => String
def sell(drink, person)
  "Here's your #{drink}."
end
```
[full code](https://gist.github.com/shockone/747ad8d06de0e195849b#file-sell-rb-L30-L38)

It works perfectly well, but let's say I want to define the `sell` method on `Person`. Now it becomes not obvious how would I check in the person is allowed to have the drink.

**Comment by waterlink** [comment]
Just like that (normal OO way):

```ruby
class Underaged
  include Contracts::Core

  Contract Drink => String
  def sell(drink)
    "Sorry, come back in #{18 - person.age} years."
  end
end

class Person
  include Contracts::Core

  Contract Drink => String
  def sell(drink)
    "Here is your drink."
  end
end
```

**Comment by shockone** [comment]
@waterlink, but now I have to create a factory to create a person depending on age.
You propose to use polymorphism, whereas the gem advertises to enable methods overloading, which I'm trying to achieve. I understand that normally languages don't dispatch based on `self`, but why not?

**Comment by nixpulvis** [comment]
Thing is you already have a `Underaged` defined in your example.

**Comment by shockone** [comment]
I do have an `Underaged` contract, but it's a constraint, not a class, which I'd use for polymorphic behaviour.

**Comment by shockone** [comment]
I'm willing to implement the feature providing we agree on the syntax and its usefulness.

**Comment by alex-fedorov** [comment]
Propose a syntax then :)

Other way: just use private method, that doesn't depend on `self`. Like this:

```ruby
class Person
  Contract Drink => String
  def sell(drink)
    _sell(self, drink)
  end

  private

  Contract Underaged, Drink => String
  def _sell(person, drink)
    "Sorry, come back in #{18 - person.age} years."
  end

  Contract Person, Drink => String
  def _sell(person, drink)
    "Here is your #{drink}!"
  end
end
```

**Comment by shockone** [comment]
I'm afraid I don't have enough knowledge about which symbols can be used. Since you chose `=>` to delimit arguments from the return value, I guess `->` is not possible. Maybe something like
```ruby
Contract Alkohol > Underaged => String
Contract Alkohol | Underaged => String
Contract Alkohol >> Underaged => String

```

**Comment by alex-fedorov** [comment]
How about:

```ruby
Contract Contracts::Receiver[Underaged], Drink => String
```

?

**Comment by shockone** [comment]
You know better, but won't it cause difficulties to distinguish whether the contract is for the first parameter or for the receiver?

**Comment by alex-fedorov** [comment]
It will cause this problem in any case.

**Comment by shockone** [comment]
I mean you chose the same separator as for the parameters list. If the separator was different, it would be easier to tell the difference.

**Comment by egonSchiele** [comment]
This does sound like a useful feature. Maybe we could do

```ruby
Contract(Underaged) Drink => String

and

Contract(Person) Drink => String
```

If we have that code, it will build the foundation for future changes, where we can pass in arbitrary data to the `Contract` function before specifying the contract itself.

**Comment by waterlink** [comment]
Unfortunately, this is not a valid ruby. To make it valid we will have to send some message after calling `Contract(x)`:

```ruby
Contract(Underaged).with Drink => String
```

**Comment by waterlink** [comment]
Another example would be to return a lambda and call it:

```ruby
Contract(Underaged).(Drink => String)
# or
Contract(Underaged)[Drink => String]
# or
Contract[Underaged].(Drink => String)
# or
Contract[Underaged][Drink => String]
```

**Comment by waterlink** [comment]
Some interesting syntax still can be used:

```ruby
Contract[Underaged].>> Drink => String
```

**Comment by waterlink** [comment]
Last one resembles Haskell version the most.

**Comment by egonSchiele** [comment]
Yeah that's what I was thinking @waterlink...`Contract(Underage)` would return a callable object. Then `Contract(Underage) foo => bar` would be like calling that function with `foo => bar`.

I'm also ok with the `>>` syntax:

```ruby
Contract Underage >> Foo => Bar
```

But I thought that might get confusing and start to look like symbol soup. @shockone once we agree on a syntax a PR for this would be great!

**Comment by waterlink** [comment]
@egonSchiele This does not work, I have tried it before. Even if `Contract(Underage)` returns callable object, ruby parser can't handle `something(args) more, args`. And callable objects (such as method, proc, block and lambda) in ruby can be called only with:

- `#call`, ie: `callable.call(args)` or `callable.(args)`
- `#[]`, ie: `callable[args]`

**Comment by waterlink** [comment]
@egonSchiele What about using a word instead of a symbol, ie: `Contract(Underage).with Foo => Bar`? Or any other short word, which means the same: "Underage with args: Foo and returns Bar".

**Comment by egonSchiele** [comment]
Yeah, I like `with` too, but would we be able to keep existing syntax? Like `Contract Num => Num`?

**Comment by waterlink** [comment]
Not a problem. We need to detect on fact if there is a hash in arguments to `Contract` or not. If yes: it is old syntax and we just instantiate contract object right from the bet. If no: we return contract factory object, that have `#with` or `#>>` method.

**Comment by shockone** [comment]
Just trying to figure out something that looks nice:
```ruby
Contract Underaged => String { Alkohol } # Not sure if the block will be passed to String or Contract, though.
ContractWithSelf Alkohol, Underaged => String
SelfContract Alkohol; Contract Underaged => String
```

**Comment by shockone** [comment]
Is this one possible?
```ruby
Contract Alkohol => Underaged => String
```

**Comment by alex-fedorov** [comment]
It is if you wrap everything in parenthesis and use .>> instead of => ;-)

But on the other issue we agreed that we don't want to expose Ruby people
to currying concept here..

**Comment by waterlink** [comment]
Ok. I'm definitely up for `Contract[Underaged].with Drink => String`

**Comment by sfcgeorge** [comment]
It seems we're coming up with the same kind of syntax suggestions for 3 problems, this one plus:

Namespacing: https://github.com/egonSchiele/contracts.ruby/issues/157#issuecomment-111726998
```ruby
Contract {[ num, maybe[num] => String ]}
```
Arguments: https://github.com/egonSchiele/contracts.ruby/issues/113#issuecomment-87329450
```ruby
Contract(:a, :b) { [ArrayOf[a], Func[a => b] => ArrayOf[b]] }
```

The three issues should probably be thought of together to prevent clashes and enhance simplicity.

I wonder if the arguments idea could solve this too somehow. 

```ruby
Underaged = lambda { |person| person.age < 18 }
# Have magic argument symbol :this that is always self
Contract(:this) { [Self[Underaged[this]], Drink => String] }
# But seems redundant as Self contract is needed to identify 
# the first in list as self contract rather than param contract anyway
Contract { [Self[Underaged], Drink => String] }
# Or make Contract able to take either: 
# a symbol for named args behaviour, lambda for self validation:  
Contract(Underaged) { [Drink => String] }
# You can use both then:
Contract(OldEnough, :drink) { [Alcohol[:drink] => Vomit[:drink]] }
# It makes sense to limit to only 1 self contract
# and any number of argument contracts.
# So maybe only the first can be a self contract as above,
# or use array for argument contracts:
Contract(OldEnough, [:drink, :bill]) { [Tab[:bill], Alcohol[:drink] => [Overdraft[:bill], Vomit[:drink]]] }
# Perhaps the array style could be optional
```

Or I may have completely missed what is going on, these things are a bit mind bending.

**Comment by waterlink** [comment]
I think you are correct.

Though I wanted to see this happening in more incremental way. (of course bearing in mind that all 3 should be implemented eventually and not making silly mistakes that will prevent from fixing other 2).

**Comment by sfcgeorge** [comment]
I keep forgetting about invariants but it just occurred to me that they use a block syntax already, so my vote is on the bracebox syntax for any fancy functional features added.

ruby```
invariant(:day) { 1 <= day && day <= 31 }
```
